File: code/demo/static/demo4.js
/* 시연 ④ v3 — 좌측 14건 목록 + 우측 단일 카드.
 * 흐름:
 *  - 진입 → GET /eval-set → 좌측 목록 채움 (정답 정보만)
 *  - 좌측 클릭 → 우측에 그 건의 정답 carded 표시 (LLM 영역은 "실행 대기" 표시)
 *  - "▶ 선택 1건 실행" → POST /run-one → 우측 LLM 영역 채워짐 + 좌측 행 색 라벨
 *  - "⏵ 14건 전체 실행" → SSE /run → 모든 행 차례로 채움 */

(() => {
  "use strict";

  const VERDICT_LABEL = {
    pass:    { txt: "✓", cls: "pass",    big: "✓ 키워드 통과" },
    partial: { txt: "⚠", cls: "partial", big: "⚠ 부분 일치" },
    fail:    { txt: "✗", cls: "fail",    big: "✗ 실패" },
  };
  const EXEC_LABEL = {
    first_try: "✓ 1회 성공",
    retry:     "🔁 자가 수정",
    fallback:  "🔄 Fallback",
    failure:   "❌ 거절·실패",
  };

  const state = {
    cases: [],
    results: {},   // id → case_done 데이터
    selectedId: null,
    inProgress: false,
    es: null,
    counts: { pass: 0, partial: 0, fail: 0, first_try: 0, retry: 0, fallback: 0, failure: 0 },
    done: 0,
  };

  const els = {
    list: document.getElementById("case-list"),
    runOneBtn: document.getElementById("run-one-btn"),
    runAllBtn: document.getElementById("run-all-btn"),
    placeholder: document.getElementById("detail-placeholder"),
    body: document.getElementById("detail-body"),
    sideSummary: document.getElementById("side-summary"),
    kwPass: document.getElementById("kw-pass"),
    kwPartial: document.getElementById("kw-partial"),
    kwFail: document.getElementById("kw-fail"),
    exFirst: document.getElementById("ex-first"),
    exRetry: document.getElementById("ex-retry"),
    exFallback: document.getElementById("ex-fallback"),
    exFailure: document.getElementById("ex-failure"),
    miniFill: document.getElementById("mini-fill"),
    miniLabel: document.getElementById("mini-label"),
    modal: document.getElementById("graph-modal"),
    modalSub: document.getElementById("graph-modal-sub"),
    modalBody: document.getElementById("graph-modal-body"),
  };

  function escapeHTML(s) {
    return String(s ?? "").replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
  }

  function caseItemHTML(c) {
    return `
      <li class="case-item" data-id="${escapeHTML(c.id)}">
        <span class="case-id">${escapeHTML(c.id)}</span>
        <span class="case-diff">${c.difficulty}</span>
        <span class="case-q" title="${escapeHTML(c.question)}">${escapeHTML(c.question)}</span>
        <span class="case-verdict empty" data-role="verdict">·</span>
      </li>`;
  }

  function renderCaseList() {
    els.list.innerHTML = state.cases.map(caseItemHTML).join("");
    els.list.querySelectorAll(".case-item").forEach((li) => {
      li.addEventListener("click", () => selectCase(li.dataset.id));
    });
  }

  function selectCase(id) {
    state.selectedId = id;
    els.list.querySelectorAll(".case-item").forEach((li) => {
      li.classList.toggle("is-active", li.dataset.id === id);
    });
    els.runOneBtn.disabled = false;
    renderDetail();
  }

  function chipExpected(kw, hits) {
    if (!hits) return `<span class="kw-chip expected">${escapeHTML(kw)}</span>`;
    const cls = hits.includes(kw) ? "hit" : "miss";
    const icon = cls === "hit" ? "✓" : "✗";
    return `<span class="kw-chip ${cls}">${icon} ${escapeHTML(kw)}</span>`;
  }

  function chipForbidden(kw, violations) {
    if (!violations) return `<span class="kw-chip forbidden">${escapeHTML(kw)}</span>`;
    const cls = violations.includes(kw) ? "violation" : "clean";
    const icon = cls === "violation" ? "⚠" : "✓";
    return `<span class="kw-chip ${cls}">${icon} ${escapeHTML(kw)}</span>`;
  }

  function renderFacts(result) {
    if (!result || result.length === 0) return `<div class="facts-empty">조회 결과 없음</div>`;
    return result.slice(0, 5).map((row) => {
      const lines = Object.entries(row).map(
        ([k, v]) => `<div class="facts-row"><span class="fact-k">${escapeHTML(k)}:</span><span class="fact-v">${escapeHTML(typeof v === "number" ? String(v) : String(v))}</span></div>`
      ).join("");
      return `<div style="border-bottom: 1px dashed var(--border); padding: 4px 0;">${lines}</div>`;
    }).join("") + (result.length > 5 ? `<div class="facts-empty">… 외 ${result.length - 5}건</div>` : "");
  }

  function renderDetail() {
    if (!state.selectedId) return;
    const c = state.cases.find((x) => x.id === state.selectedId);
    if (!c) return;
    const r = state.results[c.id];
    els.placeholder.hidden = true;
    els.body.hidden = false;

    // 정답 영역 + (결과 있으면 LLM 영역)
    const expChips = (c.golden.expected_keywords || []).map((k) => chipExpected(k, r?.matches?.expected_hits)).join("");
    const forChips = (c.golden.forbidden_keywords || []).map((k) => chipForbidden(k, r?.matches?.forbidden_violations)).join("");

    const verdictKind = r?.matches?.keyword_verdict;
    const verdictLabel = verdictKind ? VERDICT_LABEL[verdictKind] : null;
    const verdictHTML = verdictLabel
      ? `<span class="compare-verdict ${verdictLabel.cls}">${verdictLabel.big}</span>`
      : state.inProgress
        ? `<span class="compare-verdict running">⏳ 실행 중</span>`
        : `<span class="compare-verdict pending">대기</span>`;

    let genHTML;
    if (r) {
      const g = r.generated || {};
      const x = r.execution || {};
      const advisoryChips = (g.advisory || []).map((a) => {
        let ac = "alert";
        if (a.includes("MOCK")) ac = "mock";
        else if (a.includes("안전") || a.includes("⛔")) ac = "hazard";
        return `<span class="advisory-chip ${ac}">${escapeHTML(a)}</span>`;
      }).join("");
      const attemptLabel = g.attempt === "fallback" ? "예비 답안 (golden)" : `시도 ${g.attempt ?? "—"}`;
      const hasGraph = (g.result_count ?? 0) > 0;
      const graphBtnHTML = hasGraph
        ? `<button type="button" class="btn-graph-view" data-role="open-graph">🔎 그래프에서 보기</button>`
        : "";
      genHTML = `
        <div class="compare-label">LLM 생성 TypeQL (${escapeHTML(attemptLabel)})</div>
        <div class="typeql-mini">${escapeHTML(g.typeql || "(생성 실패)")}</div>

        <div class="compare-label compare-label-row">
          <span>그래프에서 조회된 사실 (${g.result_count ?? 0}건)</span>
          ${graphBtnHTML}
        </div>
        <div class="facts-list">${renderFacts(g.result || [])}</div>

        <div class="compare-label">자연어 답변</div>
        <div class="compare-answer">${escapeHTML(g.answer || "")}</div>
        ${advisoryChips ? `<div class="advisory-row" style="margin-top:6px;">${advisoryChips}</div>` : ""}

        <div class="match-summary">
          <span class="match-line">키워드: <strong>${escapeHTML(r.matches?.keyword_summary || "")}</strong></span>
          <span class="summary-divider">|</span>
          <span class="match-line">실행: <strong>${escapeHTML(EXEC_LABEL[x.outcome] || x.outcome || "")} (${escapeHTML(x.summary || "")})</strong></span>
        </div>
      `;
    } else {
      genHTML = `<div style="color: var(--text-muted); font-size: 14px; padding: 24px 0;">▶ 실행 버튼을 누르면 AI가 그래프에 질의하고 답변까지 만들어 비교 결과를 표시합니다.</div>`;
    }

    els.body.innerHTML = `
      <header class="compare-head">
        <span class="compare-id">${escapeHTML(c.id)}</span>
        <span class="compare-difficulty">난이도 ${c.difficulty} · ${escapeHTML(c.category)}</span>
        ${verdictHTML}
      </header>
      <div class="compare-question">${escapeHTML(c.question)}</div>
      <div class="compare-cols">
        <div class="compare-col golden">
          <h4>정답 (Golden)</h4>
          <span class="ontology-path">${escapeHTML(c.golden.ontology_path || "(경로 없음)")}</span>
          <div class="compare-label">정답 TypeQL</div>
          <div class="typeql-mini">${escapeHTML(c.golden.typeql || "")}</div>
          <div class="compare-label">기대 키워드</div>
          <div class="keyword-chip-row">${expChips || '<span class="facts-empty">(없음)</span>'}</div>
          <div class="compare-label">금지어</div>
          <div class="keyword-chip-row">${forChips || '<span class="facts-empty">(없음)</span>'}</div>
        </div>
        <div class="compare-col generated">
          <h4>LLM 생성</h4>
          ${genHTML}
        </div>
      </div>
    `;

    const openBtn = els.body.querySelector('[data-role="open-graph"]');
    if (openBtn) {
      openBtn.addEventListener("click", () => openGraphModal(c, r));
    }
  }

  // === 그래프 시각화 모달 (B안) ===
  // result row의 dict에서 entity 식별자 후보를 뽑아 노드 헤더로 사용
  const ID_KEYS = ["id", "segment-id", "segment_id", "source-id", "source_id"];
  const NAME_KEYS = ["name", "segment-name", "segment_name", "source-name", "source_name"];

  function pickRowHeader(row, index) {
    for (const k of ID_KEYS) {
      if (row[k] !== undefined && row[k] !== null && row[k] !== "") return { key: k, value: row[k] };
    }
    for (const k of NAME_KEYS) {
      if (row[k] !== undefined && row[k] !== null && row[k] !== "") return { key: k, value: row[k] };
    }
    return { key: "row", value: `#${index + 1}` };
  }

  function inferEntityType(ontologyPath, headerKey) {
    const path = (ontologyPath || "").toLowerCase();
    const key = (headerKey || "").toLowerCase();
    if (key.startsWith("source") || path.startsWith("source")) return "Source";
    if (key.startsWith("segment") || path.startsWith("segment")) return "Segment";
    if (path.startsWith("runcontext") || path.startsWith("run-context")) return "RunContext";
    if (path.startsWith("threshold") || path.startsWith("weight")) return "ConfigSet";
    if (path.startsWith("decision")) return "Decision";
    return "Entity";
  }

  // expected keyword 중 attribute 값을 강조해야 하는 것 골라내기 ('A|B' → ['A','B'])
  function expandExpected(expected) {
    const out = [];
    for (const kw of expected || []) {
      const parts = String(kw).includes("|") ? String(kw).split("|") : [String(kw)];
      for (const p of parts) if (p) out.push(p);
    }
    return out;
  }

  function attrIsHighlighted(value, expectedFlat) {
    const v = String(value ?? "");
    if (!v) return false;
    return expectedFlat.some((kw) => v.includes(kw));
  }

  function buildAttributeChip(key, value, expectedFlat) {
    const highlight = attrIsHighlighted(value, expectedFlat);
    const cls = highlight ? "attr-chip is-match" : "attr-chip";
    const star = highlight ? `<span class="attr-mark" title="기대 키워드와 일치">✦</span>` : "";
    return `
      <li class="${cls}">
        <span class="attr-key">${escapeHTML(key)}</span>
        <span class="attr-arrow">→</span>
        <span class="attr-val">${escapeHTML(String(value ?? ""))}</span>
        ${star}
      </li>`;
  }

  function buildRowNode(row, index, expectedFlat, ontologyPath) {
    const header = pickRowHeader(row, index);
    const entityType = inferEntityType(ontologyPath, header.key);
    const attrEntries = Object.entries(row);
    const attrItems = attrEntries
      .map(([k, v]) => buildAttributeChip(k, v, expectedFlat))
      .join("");
    const headerValDisplay = header.key === "row" ? "" : ` <span class="row-node-id">${escapeHTML(String(header.value))}</span>`;
    return `
      <div class="row-node">
        <div class="row-node-head">
          <span class="row-node-badge">${escapeHTML(entityType)}</span>
          <span class="row-node-label">row ${index + 1}${headerValDisplay}</span>
        </div>
        <div class="row-node-trunk" aria-hidden="true"></div>
        <ul class="row-node-attrs">
          ${attrItems || '<li class="attr-chip is-empty">(attribute 없음)</li>'}
        </ul>
      </div>`;
  }

  function openGraphModal(c, r) {
    const g = r?.generated || {};
    const rows = Array.isArray(g.result) ? g.result : [];
    const ontologyPath = c.golden?.ontology_path || "(경로 미정)";
    const expectedFlat = expandExpected(c.golden?.expected_keywords || []);
    const hits = r?.matches?.expected_hits || [];

    const hitsHTML = (c.golden?.expected_keywords || [])
      .map((kw) => {
        const cls = hits.includes(kw) ? "kw-chip hit" : "kw-chip miss";
        const icon = hits.includes(kw) ? "✓" : "✗";
        return `<span class="${cls}">${icon} ${escapeHTML(kw)}</span>`;
      })
      .join("");

    const rowsHTML = rows.length
      ? rows.map((row, i) => buildRowNode(row, i, expectedFlat, ontologyPath)).join("")
      : `<div class="graph-empty">조회 결과가 없습니다.</div>`;

    els.modalSub.textContent = `${c.id} · ${c.question}`;
    els.modalBody.innerHTML = `
      <section class="graph-section">
        <h3 class="graph-section-title">온톨로지 경로</h3>
        <div class="ontology-path-big">${escapeHTML(ontologyPath)}</div>
      </section>

      <section class="graph-section">
        <h3 class="graph-section-title">기대 키워드 ↔ 답변 매칭</h3>
        <div class="keyword-chip-row">${hitsHTML || '<span class="facts-empty">(없음)</span>'}</div>
        <p class="graph-hint">아래 노드 중 <span class="attr-mark inline">✦</span> 마크는 기대 키워드와 일치한 attribute입니다.</p>
      </section>

      <section class="graph-section">
        <h3 class="graph-section-title">그래프 조회 결과 (${rows.length}건)</h3>
        <div class="row-nodes">${rowsHTML}</div>
      </section>
    `;

    els.modal.hidden = false;
    document.body.classList.add("modal-open");
    // 닫기 버튼/배경 바인딩 (이미 있는 요소라 한 번만)
    if (!els.modal.dataset.bound) {
      els.modal.querySelectorAll('[data-role="modal-close"]').forEach((el) => {
        el.addEventListener("click", closeGraphModal);
      });
      document.addEventListener("keydown", (e) => {
        if (e.key === "Escape" && !els.modal.hidden) closeGraphModal();
      });
      els.modal.dataset.bound = "1";
    }
  }

  function closeGraphModal() {
    els.modal.hidden = true;
    document.body.classList.remove("modal-open");
  }

  function applyVerdictToListRow(id, verdictKind) {
    const li = els.list.querySelector(`.case-item[data-id="${id}"]`);
    if (!li) return;
    li.classList.remove("is-done-pass", "is-done-partial", "is-done-fail", "is-running");
    if (verdictKind) {
      li.classList.add(`is-done-${verdictKind}`);
      const verdict = li.querySelector('[data-role="verdict"]');
      const v = VERDICT_LABEL[verdictKind];
      verdict.className = `case-verdict ${v.cls}`;
      verdict.textContent = v.txt;
    }
  }

  function setRunningRow(id) {
    const li = els.list.querySelector(`.case-item[data-id="${id}"]`);
    if (!li) return;
    li.classList.add("is-running");
    const verdict = li.querySelector('[data-role="verdict"]');
    verdict.className = "case-verdict running";
    verdict.textContent = "⏳";
  }

  function clearRunningRow(id) {
    const li = els.list.querySelector(`.case-item[data-id="${id}"]`);
    if (li) li.classList.remove("is-running");
  }

  // === API ===
  async function loadEvalSet() {
    const res = await fetch("/api/demo4/eval-set");
    const data = await res.json();
    state.cases = data.cases || [];
    renderCaseList();
  }

  async function runOne() {
    if (!state.selectedId) return;
    state.inProgress = true;
    els.runOneBtn.disabled = true;
    els.runAllBtn.disabled = true;
    setRunningRow(state.selectedId);
    renderDetail();

    try {
      const res = await fetch("/api/demo4/run-one", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ case_id: state.selectedId }),
      });
      if (!res.ok) {
        const data = await res.json().catch(() => ({}));
        throw new Error(data.detail || `HTTP ${res.status}`);
      }
      const result = await res.json();
      const wasNew = !state.results[state.selectedId];
      state.results[state.selectedId] = result;
      // 카운트 갱신 — 이미 결과 있는 건은 카운트에 미반영 (덮어쓰기만)
      if (wasNew) {
        state.counts[result.matches.keyword_verdict] += 1;
        state.counts[result.execution.outcome] += 1;
        state.done += 1;
        updateSummary();
      }
      applyVerdictToListRow(state.selectedId, result.matches.keyword_verdict);
      clearRunningRow(state.selectedId);
    } catch (err) {
      clearRunningRow(state.selectedId);
      alert(`실행 실패: ${err.message}`);
    } finally {
      state.inProgress = false;
      els.runOneBtn.disabled = !state.selectedId;
      els.runAllBtn.disabled = false;
      renderDetail();
    }
  }

  function runAll() {
    state.inProgress = true;
    els.runOneBtn.disabled = true;
    els.runAllBtn.disabled = true;
    els.runAllBtn.textContent = "평가 중…";
    // 결과·카운트 리셋
    state.results = {};
    state.counts = { pass: 0, partial: 0, fail: 0, first_try: 0, retry: 0, fallback: 0, failure: 0 };
    state.done = 0;
    updateSummary();
    els.list.querySelectorAll(".case-item").forEach((li) => {
      li.classList.remove("is-done-pass", "is-done-partial", "is-done-fail", "is-running");
      const verdict = li.querySelector('[data-role="verdict"]');
      verdict.className = "case-verdict empty";
      verdict.textContent = "·";
    });

    const es = new EventSource("/api/demo4/run");
    state.es = es;
    es.onmessage = (e) => {
      let ev;
      try { ev = JSON.parse(e.data); } catch { return; }
      if (ev.type === "case_start") {
        setRunningRow(ev.id);
      } else if (ev.type === "case_done") {
        clearRunningRow(ev.id);
        state.results[ev.id] = ev;
        const v = ev.matches?.keyword_verdict;
        const x = ev.execution?.outcome;
        if (v) state.counts[v] = (state.counts[v] || 0) + 1;
        if (x) state.counts[x] = (state.counts[x] || 0) + 1;
        state.done += 1;
        applyVerdictToListRow(ev.id, v);
        updateSummary();
        if (state.selectedId === ev.id) renderDetail();
      } else if (ev.type === "finished") {
        es.close();
        state.es = null;
        state.inProgress = false;
        els.runOneBtn.disabled = !state.selectedId;
        els.runAllBtn.disabled = false;
        els.runAllBtn.textContent = "⏵ 14건 전체 실행";
        if (ev.keyword_stats) {
          state.counts.pass = ev.keyword_stats.pass || 0;
          state.counts.partial = ev.keyword_stats.partial || 0;
          state.counts.fail = ev.keyword_stats.fail || 0;
        }
        if (ev.execution_stats) {
          Object.assign(state.counts, ev.execution_stats);
        }
        updateSummary();
      } else if (ev.type === "error") {
        es.close();
        state.es = null;
        state.inProgress = false;
        els.runOneBtn.disabled = !state.selectedId;
        els.runAllBtn.disabled = false;
        els.runAllBtn.textContent = "⏵ 14건 전체 실행";
        alert("평가 중단: " + (ev.message || "오류"));
      }
    };
    es.onerror = () => {
      if (es.readyState === EventSource.CLOSED) {
        es.close();
        state.es = null;
        if (state.inProgress) {
          state.inProgress = false;
          els.runAllBtn.disabled = false;
          els.runAllBtn.textContent = "⏵ 14건 전체 실행";
        }
      }
    };
  }

  function updateSummary() {
    els.sideSummary.hidden = state.done === 0;
    els.kwPass.textContent = state.counts.pass;
    els.kwPartial.textContent = state.counts.partial;
    els.kwFail.textContent = state.counts.fail;
    els.exFirst.textContent = state.counts.first_try;
    els.exRetry.textContent = state.counts.retry;
    els.exFallback.textContent = state.counts.fallback;
    els.exFailure.textContent = state.counts.failure;
    const total = state.cases.length;
    const pct = total > 0 ? (state.done / total) * 100 : 0;
    els.miniFill.style.width = `${pct}%`;
    els.miniLabel.textContent = `${state.done}/${total}`;
  }

  els.runOneBtn.addEventListener("click", runOne);
  els.runAllBtn.addEventListener("click", runAll);

  loadEvalSet().catch((err) => {
    els.list.innerHTML = `<li><div class="error-box"><strong>로드 실패</strong>${escapeHTML(err.message)}</div></li>`;
  });
})();

File: code/demo/api/demo4.py
"""시연 ④ API 라우터 — 평가셋 14건 정답 ↔ LLM 비교.

각 case에 대해:
  - 골든 자료: TypeQL, expected_keywords, forbidden_keywords, ontology_path
  - LLM 생성: TypeQL(마지막 시도), 그래프 조회 결과, 자연어 답변
  - 매칭: 키워드 포함 여부, 금지어 위반, 실행 성공 분류
  - 종합 판정: 키워드 기준 + 실행 기준 둘 다 표시

엔드포인트:
  GET /eval-set      14건 메타 미리보기
  GET /run           SSE 스트림 — case_start/case_done 이벤트
"""

import asyncio
import json
import sys
from collections import Counter
from collections.abc import AsyncGenerator
from pathlib import Path

from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse, StreamingResponse
from pydantic import BaseModel, Field

_LLM_DIR = Path(__file__).resolve().parents[2] / "llm"
sys.path.insert(0, str(_LLM_DIR))

from text_to_typeql import GOLDEN_SET, query_with_trail  # noqa: E402

# demo3의 자연어 답변 생성 + advisory 추출 재사용 (코드 중복 방지)
_API_DIR = Path(__file__).resolve().parent
sys.path.insert(0, str(_API_DIR))
from demo3 import _extract_advisory, _narrate  # noqa: E402

router = APIRouter()


# === 헬퍼 ===


def _cases_sorted() -> list[dict]:
    return sorted(GOLDEN_SET.values(), key=lambda c: c["id"])


def _classify_execution(trail_dict: dict) -> tuple[str, str]:
    """trail_dict → (outcome, summary). outcome ∈ {first_try, retry, fallback, failure}."""
    status = trail_dict.get("final_status")
    trail = trail_dict.get("trail") or []
    if status == "fallback_success":
        return "fallback", "예비 답안(golden TypeQL) 사용"
    if status == "failure":
        return "failure", "데이터 부재 · 의도된 거절"
    # success
    last_ok = next(
        (t for t in reversed(trail) if t.get("ok") and t.get("result_count", 0) > 0),
        None,
    )
    if not last_ok:
        return "failure", "결과 없음"
    if last_ok["attempt"] == 1:
        return "first_try", f"1회 성공 · {last_ok['result_count']}건"
    return "retry", f"{last_ok['attempt']}회 시도 후 성공 · {last_ok['result_count']}건"


def _last_attempt(trail_dict: dict) -> dict | None:
    trail = trail_dict.get("trail") or []
    if not trail:
        return None
    # 마지막 ok=True 시도 우선
    for t in reversed(trail):
        if t.get("ok") and t.get("result_count", 0) > 0:
            return t
    return trail[-1]


def _match_keywords(answer: str, expected: list[str]) -> tuple[list[str], list[str]]:
    """expected 각 단어가 answer에 포함됐는지. 'A|B' 형식은 둘 중 하나.

    Returns (hits, missing).
    """
    hits: list[str] = []
    missing: list[str] = []
    text = answer or ""
    for kw in expected or []:
        # 'A|B' OR 매칭
        alternatives = kw.split("|") if "|" in kw else [kw]
        if any(alt in text for alt in alternatives):
            hits.append(kw)
        else:
            missing.append(kw)
    return hits, missing


def _check_forbidden(answer: str, forbidden: list[str]) -> list[str]:
    text = answer or ""
    return [kw for kw in (forbidden or []) if kw in text]


def _keyword_verdict(hits: list, missing: list, violations: list) -> tuple[str, str]:
    """(pass|partial|fail, summary 문자열)."""
    if not missing and not violations:
        return "pass", f"기대 키워드 {len(hits)}개 모두 포함, 금지어 0개 위반"
    if violations:
        return "fail", f"금지어 {len(violations)}개 위반: {', '.join(violations)}"
    # missing 있고 violations 없음
    if len(hits) >= len(missing):
        return "partial", f"기대 키워드 {len(hits)}/{len(hits) + len(missing)} 포함"
    return "fail", f"기대 키워드 {len(hits)}/{len(hits) + len(missing)} 포함 — 누락 많음"


async def _process_case(case: dict, index: int = 0) -> dict:
    """단일 case 처리 → case_done 이벤트 형식의 dict 반환."""
    trail_dict = await asyncio.to_thread(query_with_trail, case["question"], case["id"])

    exec_outcome, exec_summary = _classify_execution(trail_dict)

    last_ok = next(
        (t for t in (trail_dict.get("trail") or []) if t.get("ok") and t.get("result_count", 0) > 0),
        None,
    )
    if last_ok:
        results = last_ok.get("result") or []
        try:
            answer = await asyncio.to_thread(
                _narrate, case["question"], results, trail_dict.get("fallback_used", False)
            )
        except Exception as exc:  # noqa: BLE001
            answer = f"(답변 생성 실패: {exc})"
        advisory = _extract_advisory(results)
    else:
        results = []
        answer = (
            f'"{case["question"]}" — 해당 시점·조건의 데이터가 그래프에 존재하지 '
            "않습니다. 그래서 답을 만들지 않습니다 (환각 방지)."
        )
        advisory = []

    hits, missing = _match_keywords(answer, case.get("expected_keywords", []))
    violations = _check_forbidden(answer, case.get("forbidden_keywords", []))
    kw_verdict, kw_summary = _keyword_verdict(hits, missing, violations)

    last_attempt = _last_attempt(trail_dict)

    return {
        "type": "case_done",
        "index": index,
        "id": case["id"],
        "difficulty": case["difficulty"],
        "category": case["category"],
        "question": case["question"],
        "golden": {
            "typeql": case.get("golden_typeql", ""),
            "expected_keywords": case.get("expected_keywords", []),
            "forbidden_keywords": case.get("forbidden_keywords", []),
            "ontology_path": case.get("ontology_path", ""),
        },
        "generated": {
            "typeql": last_attempt.get("typeql") if last_attempt else None,
            "attempt": last_attempt.get("attempt") if last_attempt else None,
            "result_count": last_attempt.get("result_count") if last_attempt else 0,
            "result": results,
            "answer": answer,
            "advisory": advisory,
        },
        "matches": {
            "expected_hits": hits,
            "expected_missing": missing,
            "forbidden_violations": violations,
            "keyword_verdict": kw_verdict,
            "keyword_summary": kw_summary,
        },
        "execution": {
            "outcome": exec_outcome,
            "summary": exec_summary,
            "trail_length": len(trail_dict.get("trail") or []),
        },
    }


# === 엔드포인트 ===


@router.get("/eval-set")
def list_eval_set() -> JSONResponse:
    cases = _cases_sorted()
    by_category: Counter = Counter()
    for c in cases:
        by_category[c["category"]] += 1
    return JSONResponse(
        {
            "total": len(cases),
            "cases": [
                {
                    "id": c["id"],
                    "difficulty": c["difficulty"],
                    "category": c["category"],
                    "question": c["question"],
                    "golden": {
                        "typeql": c.get("golden_typeql", ""),
                        "expected_keywords": c.get("expected_keywords", []),
                        "forbidden_keywords": c.get("forbidden_keywords", []),
                        "ontology_path": c.get("ontology_path", ""),
                    },
                }
                for c in cases
            ],
            "by_category": dict(by_category),
        }
    )


async def _eval_stream() -> AsyncGenerator[dict, None]:
    cases = _cases_sorted()
    total = len(cases)
    keyword_stats: Counter = Counter()
    exec_stats: Counter = Counter()

    yield {"type": "started", "total": total}

    for index, case in enumerate(cases):
        yield {"type": "case_start", "index": index, "id": case["id"]}
        await asyncio.sleep(0)

        result = await _process_case(case, index=index)
        exec_stats[result["execution"]["outcome"]] += 1
        keyword_stats[result["matches"]["keyword_verdict"]] += 1
        yield result

    yield {
        "type": "finished",
        "keyword_stats": {
            "pass": keyword_stats.get("pass", 0),
            "partial": keyword_stats.get("partial", 0),
            "fail": keyword_stats.get("fail", 0),
        },
        "execution_stats": {
            "first_try": exec_stats.get("first_try", 0),
            "retry": exec_stats.get("retry", 0),
            "fallback": exec_stats.get("fallback", 0),
            "failure": exec_stats.get("failure", 0),
        },
    }


async def _sse_wrap(gen: AsyncGenerator[dict, None]) -> AsyncGenerator[bytes, None]:
    try:
        async for ev in gen:
            yield f"data: {json.dumps(ev, ensure_ascii=False)}\n\n".encode()
    except asyncio.CancelledError:
        raise
    except Exception as exc:  # noqa: BLE001
        err = {"type": "error", "message": f"평가 중 오류: {exc!s}"}
        yield f"data: {json.dumps(err, ensure_ascii=False)}\n\n".encode()


@router.get("/run")
async def run_eval() -> StreamingResponse:
    headers = {"Cache-Control": "no-cache", "X-Accel-Buffering": "no"}
    return StreamingResponse(
        _sse_wrap(_eval_stream()),
        media_type="text/event-stream",
        headers=headers,
    )


class RunOneReq(BaseModel):
    case_id: str = Field(..., min_length=1, max_length=20)


@router.post("/run-one")
async def run_one(req: RunOneReq) -> dict:
    """단일 case 실행 — 비-SSE, 1회 응답. UI에서 한 건씩 인터랙티브 실행에 사용."""
    case = GOLDEN_SET.get(req.case_id)
    if not case:
        raise HTTPException(status_code=404, detail=f"알 수 없는 case_id: {req.case_id}")
    # _process_case는 case_done 이벤트 형식과 동일한 dict를 반환
    result = await _process_case(case, index=0)
    # 단일 호출 응답이라 type 필드는 제거 (클라가 SSE 이벤트로 받지 않음)
    result.pop("type", None)
    return result
